crates/bevy_ecs/src/schedule/commands.rs

```diff
@@ -0,0 +1,121 @@
+use std::marker::PhantomData;
+
+use super::{IntoSystemDescriptor, Schedule, StageLabel};
+
+#[derive(Default)]
+pub struct ScheduleCommandQueue {
+    items: Vec<Box<dyn ScheduleCommand>>
+}
+
+impl ScheduleCommandQueue {
+    pub fn push<C>(&mut self, command: C)
+    where 
+        C: ScheduleCommand,
+    {
+        self.items.push(Box::new(command));
+    }
+
+    pub fn apply(&mut self, schedule: &mut Schedule) {
+        for command in self.items.drain(..) {
+            command.write(schedule);
+        }
+    }
+
+    pub fn transfer(&mut self, queue: &mut ScheduleCommandQueue) {
+        queue.items.extend(self.items.drain(..));
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+}
+
+/// A [`Schedule`] mutation.
+pub trait ScheduleCommand: Send + Sync + 'static {
+    fn write(self: Box<Self>, schedule: &mut Schedule);
+}
+
+pub struct ScheduleCommands<'a> {
+    queue: &'a mut ScheduleCommandQueue
+}
+
+impl<'a> ScheduleCommands<'a> {
+    pub fn new(queue: &'a mut ScheduleCommandQueue) -> Self {
+        Self {
+            queue
+        }
+    }
+
+    pub fn insert_system<T: 'static, S, Params: 'static>(&mut self, system: T, stage_label: S)
+    where
+        T: IntoSystemDescriptor<Params> + Send + Sync,
+        S: StageLabel
+    {
+        self.queue.push(InsertSystem {
+            system: system,
+            stage_label: stage_label,
+            phantom: Default::default(),
+        });
+    }
+}
+
+pub struct InsertSystem<T, S, Params>
+where
+    T: IntoSystemDescriptor<Params>,
+    S: StageLabel
+{
+    pub system: T,
+    pub stage_label: S,
+    phantom: PhantomData<fn() -> Params>,
+}
+
+impl<T: 'static, S, Params: 'static> ScheduleCommand for InsertSystem<T, S, Params>
+where
+    T: IntoSystemDescriptor<Params> + Send + Sync,
+    S: StageLabel
+{
+    fn write(self: Box<Self>, schedule: &mut Schedule) {
+        schedule.add_system_to_stage(self.stage_label, self.system);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        schedule::{Schedule, ScheduleCommands, ScheduleCommandQueue, SystemStage},
+        system::Commands,
+        world::World,
+    };
+
+    #[test]
+    fn insert_system() {
+        fn sample_system(mut _commands: Commands) {
```

Comment:
I'd like this system to do something, so then we test that the new schedules actually work when run. That will help us catch initialization issues.